src/pudl/transform/ferc714.py

```diff
@@ -315,35 +361,223 @@ def _post_process(df: pd.DataFrame, table_name: str) -> pd.DataFrame:
+            .pipe(cls.convert_into_static_table_xbrl)
+            .assign(source="xbrl")
+        )
+        # CONCATED STUFF
+        df = (
+            pd.concat([csv, xbrl])
+            .reset_index(drop=True)
+            .convert_dtypes()
+            .pipe(cls.spot_fix_eia_codes, "combined")
+            .pipe(cls.ensure_eia_code_uniqueness, "combined")
+            .pipe(cls.condense_into_one_source_table)
+            .pipe(_fillna_respondent_id_ferc714_source, "csv")
+            # the xbrl version of this is fillna is not *strictly necessary*
+            # bc we are sorting the records to grab the xbrl record
+            .pipe(_fillna_respondent_id_ferc714_source, "xbrl")
+        )
+        return df
+
+    @staticmethod
+    def spot_fix_eia_codes(
+        df: pd.DataFrame, source: Literal["csv", "xbrl", "combined"]
+    ) -> pd.DataFrame:
+        """Spot fix the eia_codes."""
+        df.loc[df.eia_code == 0, "eia_code"] = pd.NA
+        suffix = "" if source == "combined" else f"_{source}"
+        # There are a few utilities that seem mappable, but missing:
+        for rid, new_code in EIA_CODE_FIXES[source].items():
+            df.loc[df[f"respondent_id_ferc714{suffix}"] == rid, "eia_code"] = new_code
+        return df
+
+    @staticmethod
+    def ensure_eia_code_uniqueness(
+        df: pd.DataFrame, source: Literal["csv", "xbrl", "combined"]
+    ) -> pd.DataFrame:
+        """Ensure there is only one unique eia_code for each respondent."""
+        df["eia_code_count"] = (
+            df.dropna(subset=["eia_code"])
+            .groupby(["respondent_id_ferc714"])[["eia_code"]]
+            .transform("nunique")
+        )
+        if not (
+            multiple_eia_codes := df[(df.eia_code_count != 1) & (df.eia_code.notnull())]
+        ).empty:
+            raise AssertionError(
+                "We expected 0 respondents with multiple different eia_code's "
+                f"reported for each respondent in {source} data, "
+                f"but we found {len(multiple_eia_codes)}"
+            )
+        return df.drop(columns=["eia_code_count"])
+
+    @staticmethod
+    def clean_eia_codes_xbrl(xbrl: pd.DataFrame) -> pd.DataFrame:
+        """Make eia_code's cleaner coming from the XBRL data.
+
+        Desired outcomes here include all respondents have only one non-null
+        eia_code and all eia_codes that are actually the respondent_id_ferc714_xbrl
+        are nulled.
+        """
+        # we expect all of these submissions to be from the last Q
+        assert all(xbrl.report_period == "Q4")
+        # first we are gonna null out all of the "EIA" codes that are really just the respondent id
+        code_is_respondent_id_mask = xbrl.eia_code.str.startswith("C") & (
+            xbrl.respondent_id_ferc714_xbrl == xbrl.eia_code
+        )
+        xbrl.loc[code_is_respondent_id_mask, "eia_code"] = pd.NA
+
+        # lets null out some of the eia_code's from XBRL that we've manually culled
```

Comment:
By manually culled do you mean that these IDs are not the actual eia_code?

Reply:
yes i added more details here